### ISSUE TITLE
Bitfinex 'Key price should be a decimal number'

### ISSUE THREAD

**Issue by joostvanhoof** [ccxt/ccxt#6447]
- OS: macOS Catalina
- Programming Language version: PHP 7.4
- CCXT version: 1.22.31
- Exchange: Bitfinex (`bitfinex` class, not `bitfinex2`)
- Method: createOrder

First of all, thanks for making this awesome library! I've been having lots of fun with it.

The issue that I have is with Bitfinex. Whenever I place an order I always get the following error:
`Key price should be a decimal number, e.g. "123.456"`. [This issue](https://github.com/ccxt/ccxt/issues/5125) is the same but it looks like it was closed without a solution.

I'm using the `bitfinex` class instead of the `bitfinex2` class because the latter doesn't support `createOrder` (yet?). I do use `bitfinex2` to get market data, and that market data I then use to set the price that eventually goes into the `createOrder` function that throws the error.

The problem is probably that I just don't understand the manual. What I'm confused about is that the manual says:

> The decimal precision counting mode is available in the exchange.precisionMode property.

For bitfinex that's defined as `SIGNIFICANT_DIGITS`.

But a few lines down the manual says

> The exchange base class contains the decimalToPrecision method to help format values to the required decimal precision with support for different rounding, counting and padding modes.

But `decimalToPrecision` is not defined in the `bitfinex` or the `bitfinex2` class. I'm guessing this is where the issue comes from, the fact that I can not use that function to convert the number to the right precision?

I'm guessing I should do something like:
```php
$exchange_class = "\\ccxt\\bitfinex";
$exchange = new $exchange_class;

$final_price = $exchange->decimalToPrecision($x, $roundingMode = ROUND, $numPrecisionDigits = null, $countingMode = DECIMAL_PLACES, $paddingMode = NO_PADDING)
```

In that function:
```php
$x = 1234.12345678
$roundingMode = ? // How / where do I find this for Bitfinex?
$numPrecisionDigits = ? // How / where do I find this for Bitfinex?
$countingMode = SIGNIFICANT_DIGITS // from the docs I understand that here I use DECIMAL_PLACES or SIGNIFICANT_DIGITS or TICK_SIZE
$paddingMode = ? // How/where do I find this for Bitfinex? 
```

Please let me know if you need more info!

**Comment by kroitor** [comment]
Hi!

> Whenever I place an order I always get the following error:
> Key price should be a decimal number, e.g. "123.456". This issue is the same but it looks like it was closed without a solution.

In order to tell the exact cause of the issue, we need a snippet of code from you to reproduce it. The snippet should be reduced to the absolute minimum (10-20 lines) containing just the exchange instantiation and your create_order call. Apart from that we also need the verbose output, including the error callstack traceback if any and all other output as well (without your keys, of course).

```Python
import ccxt
print('CCXT version:', ccxt.__version__)
exchange = ccxt.bitfinex({
    'apiKey': 'YOUR_API_KEY',
    'secret': 'YOUR_SECRET',
    'enableRateLimit': True,  # as required by the Manual
})

# load all markets and precision infos as explained in the Manual
# https://github.com/ccxt/ccxt/wiki/Manual#loading-markets
exchange.load_markets()

# enable verbose mode after loading the markets to reduce the printout
exchange.verbose = True

symbol = 'BTC/USD'  # your symbol
amount = 123.45  # your amount as float (put your values received from bitfinex2 by hand)
price = 54.321  # your price as float (put your values received from bitfinex2 manually)
side = 'buy'  # or 'sell'
order = exchange.create_order(symbol, 'limit', side, amount, price)
```

↑ What does this program say, if you edit it for your values as commented along the lines?

>  [This issue](https://github.com/ccxt/ccxt/issues/5125) is the same but it looks like it was closed without a solution.

It was closed because the minimal requirements for an issue were not satisfied after a long wait. Those requirements are explained here: https://github.com/ccxt/ccxt/wiki/FAQ

> I'm using the `bitfinex` class instead of the `bitfinex2` class because the latter doesn't support `createOrder` (yet?). 

That is correct, we will eventually replace `bitfinex` with `bitfinex2` and rename `bitfinex2` → `bitfinex`. Also, this is explained here:

- https://github.com/ccxt/ccxt/issues/6410#issuecomment-578389388

The bitfinex2 unified private API is on the way in this PR (soon):

- https://github.com/ccxt/ccxt/pull/6369

I do use `bitfinex2` to get market data, and that market data I then use to set the price that eventually goes into the `createOrder` function that throws the error.

>  I do use `bitfinex2` to get market data, and that market data I then use to set the price that eventually goes into the `createOrder` function that throws the error.

We need the code itself instead of a description of the code, hope for your understanding. Keep it as short as possible (extremely short, remove all irrelevant parts).

> The problem is probably that I just don't understand the manual. 

Looks like it. The main thing you have to understand – you don't need to format those values manually, unless you're doing something very specific. If you're using `create_order` to place the order – that method should format those values for you. The problem is that you didn't paste the code to reproduce it, therefore we don't know if it's coming from `create_order` or elsewhere.

> What I'm confused about is that the manual says:
> > The decimal precision counting mode is available in the exchange.precisionMode property.
> For bitfinex that's defined as `SIGNIFICANT_DIGITS`.
> But a few lines down the manual says
> > The exchange base class contains the decimalToPrecision method to help format values to the required decimal precision with support for different rounding, counting and padding modes.

Yes, you don't need to mind that, if you're using the Unified CCXT API. You only need that if you're using Implicit CCXT methods. The difference is also explained in the Manual. Most likely you want the Unified API.

> But `decimalToPrecision` is not defined in the `bitfinex` or the `bitfinex2` class. 

Yes, because it is defined in the base Exchange class, from which all exchanges are derived.

> I'm guessing this is where the issue comes from, the fact that I can not use that function to convert the number to the right precision?

No, because the method is actually available. The problem comes from the rest of the code that is around. Again, hard to tell the exact reason blindly, without seeing the code.

The following is probably not related directly to the cause of this issue, but to answer your questions:

> - $roundingMode = ? // How / where do I find this for Bitfinex?

Rounding mode is always TRUNCATE for amounts, ROUND for prices. So the rounding mode depends on which value you're rounding.

> $numPrecisionDigits = ? // How / where do I find this for Bitfinex?

These values are contained in the market information and should be loaded along with the other market info (market-specific), as explained in the Manual: 

- https://github.com/ccxt/ccxt/wiki/Manual#loading-markets
- https://github.com/ccxt/ccxt/wiki/Manual#precision-and-limits

> $countingMode = SIGNIFICANT_DIGITS // from the docs I understand that here I use DECIMAL_PLACES or SIGNIFICANT_DIGITS or TICK_SIZE

It's already stored in an exchange property called `exchange.precisionMode === SIGNIFICANT_DIGITS` for `bitfinex`:

- https://github.com/ccxt/ccxt/blob/master/js/bitfinex.js#L337

> $paddingMode = ? // How/where do I find this for Bitfinex? 

Assume `NO_PADDING` everywhere.

**Comment by joostvanhoof** [comment]
Thanks for the detailed reply, things are much more clear now. From your answer I also understand that it should work out of the box, right?

I'll try to provide as much info as possible for debuggin but it might take a while as I have a bot running on a server so I need to run some code locally to get the verbose output.

To start, here's the code that places the order. It's the same for all exchanges (and the other exchanges don't produce any errors) because as you'll see I instantiate the class via a variable.

I've pulled the values that went into `createOrder` from my logs and put them in the code snippet below so these were the exact values that produced the error.

Note that in a newer version of my app, which is not yet live and so it hasn't yet used the `createOrder` function, the volume and price that I use will be stored and input into `createOrder` in the format of a `DECIMAL(36,18)` instead of the `DECIMAL(16,8)` that I use in the code below.

Let me know if you need any more information.

**Code used:**
```
if($exchange_ccxt_id == 'bitfinex2')
	$exchange_ccxt_id = 'bitfinex'; // As indicated I use bitfinex2 for market data (happens before this in the app) so I switch it here for bitfinex so that I can use createOrder
 
$exchange_class = "\\ccxt\\$exchange_ccxt_id";
$exchange = new $exchange_class (array (
	'apiKey' => key here,
	'secret' => secret here,
));

try {
	$api_response = $exchange->createOrder('XVG/BTC', 'limit', 'buy', 10000.00000000, 0.00000044);
	$response = $api_response;
} catch (\Exception $e) {
	// I do some database logging here which is irrelevant
	$response = 'error';
} finally {
	return $response;
}
```

**Comment by kroitor** [comment]
>  From your answer I also understand that it should work out of the box, right?

Yes, exactly. However, there was a bug with rounding in SIGNIFICANT_DIGITS mode. And thanks for reporting the above info, it really helped locate the cause of it, so, I've uploaded a fix and it should work normally in version 1.22.33. The new build will arrive in 15 minutes. We will be happy if you report back, whether the issue is fixed on your side or not after upgrading to version 1.22.33. Thx again!
